File: 42.cpp
#include <iostream>
#include <string>
#include <vector>
#include <cctype>
#include <cassert>
using namespace std;

class UBigInt {
public:
  UBigInt() {}
  UBigInt(const string& str) {
    data.resize(str.size());
    auto p = data.rbegin();
    for (int i = 0; i < str.size(); ++i) {
      assert(isdigit(str[i]));
      *(p++) = str[i] - '0';
    }
  }
  bool IsZero() {return data.size() == 1 && data[0] == 0;}
  friend UBigInt operator+(const UBigInt& a, const UBigInt& b);
  friend UBigInt operator-(const UBigInt& a, const UBigInt& b);
  friend UBigInt operator*(const UBigInt& a, const UBigInt& b);
  friend bool operator==(const UBigInt& a, const UBigInt& b);
  friend bool operator<(const UBigInt& a, const UBigInt& b);
  friend ostream& operator<<(ostream& os, const UBigInt& num);
private:
  void DeleteLeaderZero();
  void Carry();
private:
  vector<int> data;
};

UBigInt operator+(const UBigInt& a, const UBigInt& b) {
  auto pa = a.data.begin();
  auto pb = b.data.begin();
  int carry = 0;
  UBigInt c;
  while (pa != a.data.end() || pb != b.data.end() || carry > 0) {
    if (pa != a.data.end())
      carry += *(pa++);
    if (pb != b.data.end())
      carry += *(pb++);
    c.data.push_back(carry % 10);
    carry /= 10;
  }
  if (c.data.size() == 0) c.data.push_back(0);
  return c;
}

UBigInt operator-(const UBigInt& a, const UBigInt& b) {
  // assume a >= b
  if (a < b) return b - a;
  auto pa = a.data.begin();
  auto pb = b.data.begin();
  bool borrow = false;
  UBigInt c;
  while (pa != a.data.end()) {
    int v = *(pa++) - borrow;
    if (pb != b.data.end()) v -= *(pb++);
    if (v < 0) {
      v += 10;
      borrow = true;
    } else {
      borrow = false;
    }
    c.data.push_back(v);
  }
  // delete leader 0
  c.DeleteLeaderZero();
  return c;
}

UBigInt operator*(const UBigInt& a, const UBigInt& b) {
  UBigInt c;
  c.data.resize(a.data.size() + b.data.size(), 0);
  for (int bi = 0; bi < b.data.size(); ++bi) {
    for (int ai = 0; ai < a.data.size(); ++ai) {
      int v = a.data[ai] * b.data[bi];
      c.data[ai + bi] += v;
    }
  }
  c.Carry();
  c.DeleteLeaderZero();
  return c;
}

bool operator==(const UBigInt& a, const UBigInt& b) {
  if (a.data.size() != b.data.size()) return false;
  for (auto pa = a.data.rbegin(), pb = b.data.rbegin(); pa != a.data.rend(); ++pa, ++pb) {
    if (*pa != *pb) return false; 
  }
  return true;
}

bool operator<(const UBigInt& a, const UBigInt& b) {
  if (a.data.size() != b.data.size()) return a.data.size() < b.data.size();
  for (auto pa = a.data.rbegin(), pb = b.data.rbegin(); pa != a.data.rend(); ++pa, ++pb) {
    if (*pa != *pb) return *pa < *pb;
  }
  return false;
}

ostream& operator<<(ostream& os, const UBigInt& num) {
  for (auto p = num.data.rbegin(); p != num.data.rend(); ++p) os << *p;
  return os;
}

void UBigInt::DeleteLeaderZero() {
  auto pc = data.rbegin();
  while (pc != data.rend() && *pc == 0) ++pc;
  int size = data.rend() - pc;
  if (size == 0) size = 1;
  data.resize(size);
}

void UBigInt::Carry() {
  int carry = 0;
  for (auto p = data.begin(); p != data.end(); ++p) {
    *p += carry;
    carry = *p / 10;
    *p %= 10;
  }
  while (carry > 0) {
    data.push_back(carry % 10);
    carry /= 10;
  }
}


class BigInt {
public:
  BigInt() {}
  BigInt(const string& str) {
    int i = 0;
    if (str[0] == '-') {
      minus = true;
      ++i;
    }
    else {
      minus = false;
      if (str[0] == '+') ++i; 
    }
    data = UBigInt(str.substr(i));
  }
  friend ostream& operator<<(ostream& os, const BigInt& num);
  friend BigInt operator+(const BigInt& a, const BigInt& b);
  friend BigInt operator-(const BigInt& a, const BigInt& b);
  friend BigInt operator*(const BigInt& a, const BigInt& b);
private:
  void PostProcess();
private:
  bool minus;
  UBigInt data;
};

ostream& operator<<(ostream& os, const BigInt& num) {
  if (num.minus) os << '-';
  cout << num.data;
  return os;
}

BigInt operator+(const BigInt& a, const BigInt& b) {
  BigInt c;
  if (a.minus != b.minus) {
    if (a.data < b.data) {
      c.minus = b.minus;
      c.data = b.data - a.data;
    } else {
      c.minus = a.minus;
      c.data = a.data - b.data;
    }
    return c;
  }
  c.minus = a.minus; 
  c.data = a.data + b.data;
  c.PostProcess();
  return c;
}

BigInt operator-(const BigInt& a, const BigInt& b) {
  BigInt c;
  if (a.minus != b.minus) {
    c.minus = a.minus;
    c.data = a.data + b.data;
    return c;
  }
  if (a.data < b.data) {
    c.minus = !a.minus;
    c.data = b.data - a.data;
  } else {
    c.minus = !a.minus;
    c.data = a.data - b.data;
  }
  c.PostProcess();
  return c;
}

BigInt operator*(const BigInt& a, const BigInt& b) {
  BigInt c;
  c.minus = a.minus ^ b.minus;
  c.data = a.data * b.data;
  c.PostProcess();
  return c;
}

void BigInt::PostProcess() {
  if (data.IsZero())
    minus = false;
}

const string a_str = "-80538738812075974";
const string b_str = "80435758145817515";
const string c_str = "12602123297335631";

int main() {
  BigInt a(a_str);
  BigInt b(b_str);
  BigInt c(c_str);
  BigInt wa = a * a * a;
  BigInt wb = b * b * b;
  BigInt wc = c * c * c;
  BigInt out = a * a * a + b * b * b + c * c * c;
  cout << out << endl;
  return 0;
}

File: 42.py
class UBigInt:
    def __init__(self, string=None):
        if string is None:
            self.data = []
            return
        assert isinstance(string, str),\
            TypeError('the type of input should be `str`')
        for c in string[::-1]:
            assert ord('0') <= ord(c) <= ord('9'), c
        self.data = [ord(c) - ord('0') for c in string[::-1]]

    def is_zero(self):
        return len(self.data) == 1 and self.data[0] == 0

    def __str__(self):
        return ''.join(str(v) for v in self.data[::-1])

    def __lt__(self, other):
        assert isinstance(other, UBigInt)
        if len(self.data) != len(other.data):
            return len(self.data) < len(other.data)
        for a, b in zip(self.data[::-1], other.data[::-1]):
            if a != b:
                return a < b
        return False

    def __add__(self, other):
        assert isinstance(other, UBigInt)
        out = UBigInt()
        i = j = 0
        v = 0
        while i < len(self.data) or j < len(other.data) or v > 0:
            if i < len(self.data):
                v += self.data[i]
                i += 1
            if j < len(other.data):
                v += other.data[j]
                j += 1
            out.data.append(v % 10)
            v //= 10
        return out

    def __sub__(self, other):
        assert isinstance(other, UBigInt)
        assert not self < other
        out = UBigInt()
        borrow = False
        for i, v in enumerate(self.data):
            u = v - borrow
            if i < len(other.data):
                u -= other.data[i]
            if u < 0:
                u += 10
                borrow = True
            else:
                borrow = False
            out.data.append(u)
        out._delete_leader_zero()
        return out

    def __mul__(self, other):
        out = UBigInt()
        out.data = [0] * (len(self.data) + len(other.data))
        for ai, av in enumerate(self.data):
            for bi, bv in enumerate(other.data):
                out.data[ai + bi] += av * bv
        out._carry()
        out._delete_leader_zero()
        return out

    def _carry(self):
        c = 0
        for i, v in enumerate(self.data):
            c += v
            self.data[i] = c % 10
            c //= 10
        while c > 0:
            self.data.append(c % 10)
            c //= 10

    def _delete_leader_zero(self):
        for i, v in enumerate(self.data[::-1]):
            if v != 0:
                break
        self.data = self.data[:len(self.data)-i]
        if len(self.data) == 0:
            self.data.append(0)


class BigInt:
    def __init__(self, string=None):
        if string is None:
            self.data = UBigInt()
            self.minus = False
            return
        assert isinstance(string, str),\
            TypeError('the type of input should be `str`')
        if string[0] == '-':
            self.minus = True
            string = string[1:]
        else:
            self.minus = False
        self.data = UBigInt(string)

    def __str__(self):
        if self.minus:
            return '-' + str(self.data)
        return str(self.data)

    def __add__(self, other):
        assert isinstance(other, BigInt)
        out = BigInt()
        if self.minus == other.minus:
            out.minus = self.minus
            out.data = self.data + other.data
        elif self.data < other.data:
            out.minus = other.minus
            out.data = other.data - self.data
        else:
            out.minus = self.minus
            out.data = self.data - other.data
        out._post_process()
        return out

    def __sub__(self, other):
        assert isinstance(other, BigInt)
        out = BigInt()
        if self.minus != other.minus:
            out.minus = self.minus
            out.data = self.data + other.data
        elif self.data < other.data:
            out.minus = not self.minus
            out.data = other.data - self.data
            print("!!!", out.data, other.data, self.data)
        else:
            out.minus = self.minus
            out.data = self.data - other.data
        out._post_process()
        return out

    def __mul__(self, other):
        assert isinstance(other, BigInt)
        out = BigInt()
        out.minus = self.minus ^ other.minus
        out.data = self.data * other.data
        out._post_process()
        return out

    def _post_process(self):
        if self.data.is_zero():
            self.minus = False


if __name__ == '__main__':
    a_str = "-80538738812075974"
    b_str = "80435758145817515"
    c_str = "12602123297335631"
    a = BigInt(a_str)
    b = BigInt(b_str)
    c = BigInt(c_str)
    out = a * a * a + b * b * b + c * c * c
    print(out)
